src/Runner/Middleware/HandleRunnerMiddleware.php

```diff
@@ -37,55 +35,30 @@ public function __construct(TaskHandler $taskHandler, RunnerConfig $config)
 
     public function handle(TaskRunnerContext $context, callable $next): TaskResultCollection
     {
-        return new TaskResultCollection(
-            wait(
-                /**
-                 * @return \Generator<mixed, mixed, mixed, TaskResultInterface[]>
-                 */
-                call(function () use ($context): \Generator {
-                    /**
-                     * @var \Throwable[] $errors
-                     * @var TaskResultInterface[] $results
-                     * @psalm-suppress InvalidArrayAccess
-                     * @psalm-suppress InvalidArrayOffset
-                     */
-                    [$errors, $results] = yield MultiPromise::cancelable(
-                        $this->handleTasks($context),
-                        function (TaskResultInterface $result) {
-                            return $this->config->stopOnFailure() && $result->isBlocking();
-                        }
-                    );
+        $stopOnFailure = StopOnFailure::createFromConfig($this->config);
 
-                    // Filter out canceled items:
-                    $errors = array_filter($errors, function (\Throwable $error): bool {
-                        return !$error instanceof CancelledException;
-                    });
+        $futures = array_map(
+            /** @return Future<TaskResultInterface> */
+            fn (TaskInterface $task): Future => $this->taskHandler->handle($task, $context, $stopOnFailure),
+            $context->getTasks()->toArray()
+        );
 
-                    if ($errors) {
-                        throw new MultiReasonException($errors);
+        try {
+            return new TaskResultCollection(
+                await($futures, $stopOnFailure->cancellation())
+            );
+        } catch (CancelledException $e) {
```

Comment:
Any other amp related exceptions I should be worried about?
There might be a shortcut to collect all finished results somewhere in amp instead of doing my own reduce?